File: OgreWrapper/src/OgreEngineArchive.cpp
#include "StdAfx.h"
#include "..\include\OgreEngineArchive.h"
#include "MarshalUtils.h"
#include "OgreManagedStream.h"

namespace OgreWrapper
{

OgreEngineArchive::OgreEngineArchive(const Ogre::String& name, const Ogre::String& archType)
:Ogre::Archive(name, archType),
archive(nullptr),
baseName(MarshalUtils::convertString(name))
{

}

OgreEngineArchive::~OgreEngineArchive()
{
	unload();
}

void OgreEngineArchive::load()
{
	archive = Engine::VirtualFileSystem::Instance;
}

void OgreEngineArchive::unload()
{
	archive = nullptr;
}

Ogre::DataStreamPtr OgreEngineArchive::open(const Ogre::String& filename, bool readOnly) const
{
	System::String^ file = MarshalUtils::convertString(filename);
	if(!archive->exists(file))
	{
		file = baseName + "/" + file;
	}
	return Ogre::DataStreamPtr(OGRE_NEW OgreManagedStream(filename, archive->openStream(file, Engine::Resources::FileMode::Open, Engine::Resources::FileAccess::Read)));
}

Ogre::StringVectorPtr OgreEngineArchive::list(bool recursive, bool dirs)
{
	Ogre::StringVectorPtr ret = Ogre::StringVectorPtr(OGRE_NEW_T(Ogre::StringVector, Ogre::MEMCATEGORY_GENERAL)(), Ogre::SPFM_DELETE_T);
	cli::array<String^>^ files;
	if(dirs)
	{
		files = archive->listDirectories(baseName, recursive);
	}
	else
	{
		files = archive->listFiles(baseName, recursive);
	}
	for each(String^ file in files)
	{
		ret->push_back(MarshalUtils::convertString(file));
	}
	return ret;
}

Ogre::FileInfoListPtr OgreEngineArchive::listFileInfo(bool recursive, bool dirs)
{
	Ogre::FileInfoList* fil = OGRE_NEW_T(Ogre::FileInfoList, Ogre::MEMCATEGORY_GENERAL)();
	cli::array<String^>^ files;
	if(dirs)
	{
		files = archive->listDirectories(baseName, recursive);
		for each(String^ file in files)
		{
			Ogre::FileInfo ogreInfo;
			ogreInfo.archive = this;
			ogreInfo.compressedSize = -1;
			ogreInfo.uncompressedSize = -1;
			ogreInfo.basename = MarshalUtils::convertString(System::IO::Path::GetFileName(file));
			ogreInfo.filename = MarshalUtils::convertString(file);
			ogreInfo.path = MarshalUtils::convertString(System::IO::Path::GetDirectoryName(file));
			fil->push_back(ogreInfo);
		}
	}
	else
	{
		files = archive->listFiles(baseName, recursive);
		for each(String^ file in files)
		{
			Engine::VirtualFileInfo^ archiveInfo = archive->getFileInfo(file);
			Ogre::FileInfo ogreInfo;
			ogreInfo.archive = this;
			ogreInfo.compressedSize = archiveInfo->CompressedSize;
			ogreInfo.uncompressedSize = archiveInfo->UncompressedSize;
			ogreInfo.basename = MarshalUtils::convertString(archiveInfo->Name);
			ogreInfo.filename = MarshalUtils::convertString(archiveInfo->FullName);
			ogreInfo.path = MarshalUtils::convertString(archiveInfo->DirectoryName);
			fil->push_back(ogreInfo);
		}
	}
	return Ogre::FileInfoListPtr(fil, Ogre::SPFM_DELETE_T);
}

Ogre::StringVectorPtr OgreEngineArchive::find(const Ogre::String& pattern, bool recursive, bool dirs)
{
	Ogre::StringVectorPtr ret = Ogre::StringVectorPtr(OGRE_NEW_T(Ogre::StringVector, Ogre::MEMCATEGORY_GENERAL)(), Ogre::SPFM_DELETE_T);
	cli::array<String^>^ files;
	if(dirs)
	{
		files = archive->listDirectories(baseName, MarshalUtils::convertString(pattern), recursive);
	}
	else
	{
		files = archive->listFiles(baseName, MarshalUtils::convertString(pattern), recursive);
	}
	for each(String^ file in files)
	{
		ret->push_back(MarshalUtils::convertString(file));
	}
	return ret;
}

Ogre::FileInfoListPtr OgreEngineArchive::findFileInfo(const Ogre::String& pattern, bool recursive, bool dirs)
{
	Ogre::FileInfoList* fil = OGRE_NEW_T(Ogre::FileInfoList, Ogre::MEMCATEGORY_GENERAL)();
	cli::array<String^>^ files;
	if(dirs)
	{
		files = archive->listDirectories(baseName, MarshalUtils::convertString(pattern), recursive);
		for each(String^ file in files)
		{
			Ogre::FileInfo ogreInfo;
			ogreInfo.archive = this;
			ogreInfo.compressedSize = -1;
			ogreInfo.uncompressedSize = -1;
			ogreInfo.basename = MarshalUtils::convertString(System::IO::Path::GetFileName(file));
			ogreInfo.filename = MarshalUtils::convertString(file);
			ogreInfo.path = MarshalUtils::convertString(System::IO::Path::GetDirectoryName(file));
			fil->push_back(ogreInfo);
		}
	}
	else
	{
		files = archive->listFiles(baseName, MarshalUtils::convertString(pattern), recursive);
		for each(String^ file in files)
		{
			Engine::VirtualFileInfo^ archiveInfo = archive->getFileInfo(file);
			Ogre::FileInfo ogreInfo;
			ogreInfo.archive = this;
			ogreInfo.compressedSize = archiveInfo->CompressedSize;
			ogreInfo.uncompressedSize = archiveInfo->UncompressedSize;
			ogreInfo.basename = MarshalUtils::convertString(archiveInfo->Name);
			ogreInfo.filename = MarshalUtils::convertString(archiveInfo->FullName);
			ogreInfo.path = MarshalUtils::convertString(archiveInfo->DirectoryName);
			fil->push_back(ogreInfo);
		}
	}
	return Ogre::FileInfoListPtr(fil, Ogre::SPFM_DELETE_T);
}

bool OgreEngineArchive::exists(const Ogre::String& filename)
{
	return archive->exists(baseName + "/" + MarshalUtils::convertString(filename));
}

time_t OgreEngineArchive::getModifiedTime(const Ogre::String& filename)
{
	return 0;
}

}

File: OgrePlugin/Wrapper/CustomArchives/EngineArchive/OgreEngineArchive.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using Engine;

namespace OgreWrapper
{
    class OgreEngineArchive : OgreManagedArchive
    {
        private String baseName;
        private VirtualFileSystem vfs;

        public OgreEngineArchive(String name, String archType)
            :base(name, archType)
        {
            baseName = name;
        }

        protected override void load()
        {
            vfs = VirtualFileSystem.Instance;
        }

        protected override void unload()
        {
            vfs = null;
        }

        protected override Stream doOpen(string filename)
        {
            if (!vfs.exists(filename))
            {
                filename = baseName + "/" + filename;
            }
            return vfs.openStream(filename, Engine.Resources.FileMode.Open, Engine.Resources.FileAccess.Read);
        }

        protected override void doList(bool recursive, bool dirs, IntPtr ogreStringVector)
        {
            String[] files;
	        if(dirs)
	        {
		        files = vfs.listDirectories(baseName, recursive);
	        }
	        else
	        {
		        files = vfs.listFiles(baseName, recursive);
	        }
	        foreach(String file in files)
	        {
                OgreStringVector_push_back(ogreStringVector, file);
	        }
        }

        protected override void doListFileInfo(bool recursive, bool dirs, IntPtr ogreFileList, IntPtr archive)
        {
            String[] files;
            if(dirs)
	        {
                files = vfs.listDirectories(baseName, recursive);
		        foreach(String file in files)
		        {
                    OgreFileInfoList_push_back(ogreFileList, archive, new IntPtr(-1), new IntPtr(-1), Path.GetFileName(file), file, Path.GetDirectoryName(file));
		        }
	        }
	        else
	        {
                files = vfs.listFiles(baseName, recursive);
		        foreach(String file in files)
		        {
                    pushFixedFileInfo(file, ogreFileList, archive);
		        }
	        }
        }

        protected override void dofind(string pattern, bool recursive, bool dirs, IntPtr ogreStringVector)
        {
            String[] files;
            if(dirs)
	        {
		        files = vfs.listDirectories(baseName, pattern, recursive);
	        }
	        else
	        {
                files = vfs.listFiles(baseName, pattern, recursive);
	        }
	        foreach(String file in files)
	        {
		        OgreStringVector_push_back(ogreStringVector, file);
	        }
        }

        protected override void dofindFileInfo(string pattern, bool recursive, bool dirs, IntPtr ogreFileList, IntPtr archive)
        {
            String[] files;
            if (dirs)
            {
                files = vfs.listDirectories(baseName, pattern, recursive);
                foreach (String file in files)
                {
                    OgreFileInfoList_push_back(ogreFileList, archive, new IntPtr(-1), new IntPtr(-1), Path.GetFileName(file), file, Path.GetDirectoryName(file));
                }
            }
            else
            {
                files = vfs.listFiles(baseName, pattern, recursive);
                foreach (String file in files)
                {
                    pushFixedFileInfo(file, ogreFileList, archive);
                }
            }
        }

        protected void pushFixedFileInfo(String file, IntPtr ogreFileList, IntPtr archive)
        {
            VirtualFileInfo archiveInfo = vfs.getFileInfo(file);
            String fixedFilename = archiveInfo.FullName.Replace(baseName, "");
            if(fixedFilename.StartsWith("/"))
            {
                fixedFilename = fixedFilename.Substring(1);
            }
            String fixedPath = archiveInfo.DirectoryName.Replace(baseName, "");
            if(fixedPath.StartsWith("/"))
            {
                fixedPath = fixedPath.Substring(1);
            }
            OgreFileInfoList_push_back(ogreFileList, archive, new IntPtr(archiveInfo.CompressedSize), new IntPtr(archiveInfo.UncompressedSize), archiveInfo.Name, fixedFilename, fixedPath);
        }

        protected override bool exists(string filename)
        {
            return vfs.exists(baseName + "/" + filename);
        }
    }
}
